src/dbnode/storage/index.go

```diff
@@ -2098,13 +2095,90 @@ func (i *nsIndex) CleanupExpiredFileSets(t time.Time) error {
 	return i.deleteFilesFn(filesets)
 }
 
-func (i *nsIndex) CleanupDuplicateFileSets() error {
+type corruptedInfoFile struct {
+	IndexVolumeType idxpersist.IndexVolumeType
+	File            fs.ReadIndexInfoFileResult
+}
+
+func (i *nsIndex) CleanupCorruptedFileSets() error {
+	/*
+	   Corrupted index filesets can be safely cleaned up if its not
+	   the latest volume index per index volume type/block start combo.
+
+	   We are guaranteed not to be actively writing to an index fileset once
+	   we're already writing to later volume indices.
+	*/
 	fsOpts := i.opts.CommitLogOptions().FilesystemOptions()
-	infoFiles := i.readIndexInfoFilesFn(
-		fsOpts.FilePathPrefix(),
-		i.nsMetadata.ID(),
-		fsOpts.InfoReaderBufferSize(),
+	infoFiles := i.readIndexInfoFilesFn(fs.ReadIndexInfoFilesOptions{
+		FilePathPrefix:   fsOpts.FilePathPrefix(),
+		Namespace:        i.nsMetadata.ID(),
+		ReaderBufferSize: fsOpts.InfoReaderBufferSize(),
+		IncludeCorrupted: true,
+	})
+
+	var (
+		toDelete                      []string
+		corrupted                     []corruptedInfoFile
+		latestBlockStart              int64
+		latestVolumeIndexByVolumeType = make(map[idxpersist.IndexVolumeType]int)
 	)
+	// NB: Info files should be ordered by block start.
+	for _, file := range infoFiles {
+		// This intentionally skips the latest block start as that's the active block.
+		if file.Info.BlockStart > latestBlockStart {
```

Comment:
But for cold flush (or computed namespace), active block may not necessarily be the latest one?

Reply:
Hmm maybe I should remove this comment - it's a little confusing, it's fine even if we iterate over the active block or any block for that matter that we're actively writing to (we check against latest vol idx). 

I had the comment for why we're not running the cleanup logic for all blocks. Would need to wrap it in a fn and run it outside of the loop to cover all blocks. But running against the active block is essentially a no-op so I put the comment in.